iris/man/man1/iris-readStdinUpdateShmWriteStdout.1
.TH iris-readStdinUpdateShmWriteStdout 1 "August 2011"

.SH NAME
iris-readStdinUpdateShmWriteStdout

.SH SYNOPSIS 

.B iris-readStdinUpdateShmWriteStdout [ \fI--usleep t\fR] [\fI--noread\fR] [\fI--nowrite\fR] \fIshmName\fR

.SH OPTIONS

.IP "\fI--usleep t\fR"
specifies how many microseconds to sleep at the end of each loop.  If not specified,
iris::GetUsleep() is used.

.IP "\fI--noread\fR"
specifies that stdin will not be read

.IP "\fI--nowrite\fR"
specifies that stdout will not be written

.TP 
\fIshmName\fR
.br
the name of an existing DTK shared memory file

.SH DESCRIPTION
\fBiris-readStdinUpdateShmWriteStdout\fR loops, and every loop does the
 following:

.IP "1)"
reads the DTK shared memory file \fIshmName\fR, and if it has changed from
the previous iteration, and \fI--nowrite\fR was not specified, writes the binary
contents of the shared memory file to standard output.

.IP "2"
if the shared memory has not changed, and \fI--noread\fR was not specified,
stdin is read, and if the data read are different than the contents of the
shared memory file, the data read are written to the shared memory file.

.P
It is hoped that \fBiris-readStdinUpdateShmWriteStdout\fR can be used to
transmit DTK shared memory data across ssh tunnels, but this has not actually
been tested yet.

.SH EXAMPLE

.IP "in one window type:"
.br
dtk-caveDeviceSimulator & hev-writeFloats -c 6 foo 0 ; hev-readFloats -i foo

.IP "in another window type:"
iris-readStdinUpdateShmWriteStdout --noread head | iris-readStdinUpdateShmWriteStdout --nowrite foo

.P 
move the sliders for the head and see if the first window updates to match the sliders

.SH NOTE

All reads and writes are non-blocking. The first time DTK shared memory is
read, it is written to stdout if \fI--nowrite\fR was not specified.

.SH BUGS

The name is sort of klunky- any suggestions?

.SH AUTHOR

.PP
John Kelso, kelso@nist.gov
.br
NIST High Performance Computing and Visualization Group
.br
August 2011
.br
\fBhttp://www.nist.gov/mcsd/hpcvg/\fR
 


iris/src/program/iris-readStdinUpdateShmWriteStdout.cpp
/*

  an existing DTK shared memory file is passed in
  
  in a loop:

  1) compares the current shared memory file to its contents the last loop,
     and if they are different, writes the new contents as binary data to
     stdout

  2) reads binary data from stdin the size of the shared memory file,
     compares it to a DTK shared memory file, and if it different updates
     the shared memory file and the stored contents from the last frame

  other options to control the loop speed, nor read or not write

  dtkmsg debug option to write the data to stderr

 */

#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <signal.h>
#include <dtk.h>
#include <iris/Utils.h>

bool running = true ;
////////////////////////////////////////////////////////////////////////
void cleanup(int)
{
    running = false ;
    iris::Signal(SIG_DFL) ;
}
////////////////////////////////////////////////////////////////////////
void usage()
{
    fprintf(stderr,"usage: iris-readStdinUpdateShmWriteStdout [--usleep t] [--noread] [--nowrite] fileName\n") ;
}

////////////////////////////////////////////////////////////////////////
int main(int argc, char **argv)
{
    // send messages to stderr
    dtkMsg.setFile(stderr) ;

    if (argc < 2 || argc > 6)
    {
	usage() ;
	return 1 ;
    }

    // call this when exiting
    iris::Signal(cleanup) ;

    int ticks = iris::GetUsleep() ;
    bool noread = false ;
    bool nowrite = false ;

    int c = 1 ;
    while (c<argc && argv[c][0] == '-' && argv[c][1] == '-')
    {
	if (iris::IsSubstring("--usleep",argv[c],4))
	{
	    c++ ;
	    if (c<argc && !iris::StringToInt(argv[c],&ticks)) 
	    {
		dtkMsg.add(DTKMSG_ERROR, "iris-readStdinUpdateShmWriteStdout: invalid usleep value\n") ;
		return 1 ; 
	    }
	    c++ ;
	}
	else if (iris::IsSubstring("--noread",argv[c],5))
	{
	    c++ ;
	    noread = true ;
	}
	else if (iris::IsSubstring("--nowrite",argv[c],5))
	{
	    c++ ;
	    nowrite = true ;
	}
	else
	{
	    usage() ;
	    return 1 ;
	}
    }

    if (c+1 != argc)
    {
	usage() ;
	return 1 ;
    }

    char* shmName = argv[c] ;

    dtkSharedMem* shm = new dtkSharedMem(shmName) ;
    if (shm->isInvalid()) return 1 ;
    
    size_t shmSize = shm->getSize() ;
    
    dtkMsg.add(DTKMSG_INFO, "iris-readStdinUpdateShmWriteStdout: using shared memory file \"%s\", size %d\n",shmName, shmSize) ;
    
    // use to see if shared memory changed
    char lastShmData[shmSize] ;

    // bool always write the first one
    bool first = true ;

    // reads in shm data
    char shmData[shmSize] ;
    
    // reads in stdin
    char inData[shmSize] ;

    // make stdin read non blocking, and don't write stdout until it's possible
    fd_set rfds;
    fd_set wfds;
    // no wait, just poll
    struct timespec tv;
    int retval ;

    while (running)
    {
	
	shm->read(shmData) ;
	if (first || memcmp(lastShmData,shmData,shmSize))
	{
	    first = false ;
	    if (dtkMsg.isSeverity(DTKMSG_DEBUG))
	    {
		dtkMsg.add(DTKMSG_DEBUG, "iris-readStdinUpdateShmWriteStdout: shared memory file \"%s\" externally updated\n", shmName) ;
	    }
	    memcpy(lastShmData,shmData,shmSize) ;
	    if (!nowrite)
	    {
		FD_ZERO(&wfds);
		FD_SET(1, &wfds);
		tv.tv_sec = 0 ;
		tv.tv_nsec = 0 ;
		retval = pselect(2, NULL, &wfds, NULL, &tv, NULL);
		if (retval>0)
		{
		    if (fwrite(shmData, shmSize, 1, stdout) != 1) dtkMsg.add(DTKMSG_DEBUG, "iris-readStdinUpdateShmWriteStdout: error writing %d bytes of data to stdout\n",shmSize) ;
		    fflush(stdout) ;
		}
		else if (retval < 0)
		{
		    dtkMsg.add(DTKMSG_ERROR, "iris-readStdinUpdateShmWriteStdout: %s\n",strerror_r(errno,NULL,0)) ;
		}
	    }
	}
	else if (!noread)
	{
	    FD_ZERO(&rfds);
	    FD_SET(0, &rfds);
	    tv.tv_sec = 0 ;
	    tv.tv_nsec = 0 ;
	    retval = pselect(1, &rfds, NULL, NULL, &tv, NULL);
	    if (retval > 0)
	    {
		if (fread(inData, shmSize, 1, stdin)<0) dtkMsg.add(DTKMSG_DEBUG, "iris-readStdinUpdateShmWriteStdout: error reading %d bytes of data from stdin\n",shmSize) ;
		else if (memcmp(inData,shmData,shmSize)) 
		{
		    shm->write(inData) ;
		    memcpy(lastShmData,shmData,shmSize) ;
		    if (dtkMsg.isSeverity(DTKMSG_DEBUG))
		    {
			dtkMsg.add(DTKMSG_DEBUG, "iris-readStdinUpdateShmWriteStdout: shared memory file \"%s\" updated from stdin\n", shmName) ;
		    }
		}
	    }
	    else if (retval < 0)
	    {
		dtkMsg.add(DTKMSG_ERROR, "iris-readStdinUpdateShmWriteStdout: %s\n",strerror_r(errno,NULL,0)) ;
	    }
	}
	
	

	usleep(ticks) ;
    }
    return 0 ;
}
